lib/avro_turf.rb

```diff
@@ -43,11 +45,34 @@ def decode(encoded_data, schema_name: nil, namespace: @namespace)
     schema = schema_name && resolve_schema(schema_name, namespace)
     reader = Avro::IO::DatumReader.new(nil, schema)
     dr = Avro::DataFile::Reader.new(io, reader)
+
+    # If no reader schema is specified, we'll try to find one matching the name
+    # of the writer's schema.
+    if schema_name.nil?
+      schema = load_readers_schema(reader.writers_schema)
+      reader.readers_schema = schema if schema
+    end
+
     dr.first
   end
 
   private
 
+  # Returns the reader schema matching the writer's schema if one is locally
+  # available, or nil if there's no matching reader schema.
+  def load_readers_schema(writers_schema)
+    return nil unless writers_schema.respond_to?(:fullname)
+
+    schema_name = writers_schema.fullname
+
+    begin
+      resolve_schema(schema_name)
+    rescue SchemaNotFoundError
+      @logger.warn "Could not find schema `#{schema_name}' locally; using writer's schema instead"
```

Comment:
The string "using writer's schema instead" seems inappropriate to me- this code doesn't know what schema the caller will use.  It's easy to imagine someone refactoring the decode function to have some other logic, but not noticing this message.

I'd suggest that either this function should do the full resolution, and do error logging, or it should not log the "using writer's schema instead" part of the error (the caller should log that portion, since the caller controls that.)  For example, I could imagine a function like `def load_best_schema(writers_schema, readers_schema)` which properly handles fallback.

I don't intend to dictate the design, but I find the current naming and error reporting pretty confusing.